homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,593 @@
+                run_callback_threadsafe(
+                    self.hass.loop,
+                    async_active_zone,
+                    self.hass,
+                    device.location["latitude"],
+                    device.location["longitude"],
+                ).result
+            )
+
+            if currentzone is not None:
+                intervals[device.name] = self._max_interval
+                continue
+
+            zones = (
+                self.hass.states.get(entity_id)
+                for entity_id in sorted(self.hass.states.entity_ids("zone"))
+            )
+
+            distances = []
+            for zone_state in zones:
+                zone_state_lat = zone_state.attributes["latitude"]
+                zone_state_long = zone_state.attributes["longitude"]
+                zone_distance = distance(
+                    device.location["latitude"],
+                    device.location["longitude"],
+                    zone_state_lat,
+                    zone_state_long,
+                )
+                distances.append(round(zone_distance / 1000, 1))
+
+            if distances:
+                mindistance = min(distances)
+            else:
+                continue
+
+            # Calculate out how long it would take for the device to drive
+            # to the nearest zone at 120 km/h:
+            interval = round(mindistance / 2, 0)
+
+            # Never poll more than once per minute
+            interval = max(interval, 1)
+
+            if interval > 180:
+                # Three hour drive?
+                # This is far enough that they might be flying
+                interval = self._max_interval
+
+            if (
+                device.battery_level is not None
+                and device.battery_level <= 33
+                and mindistance > 3
+            ):
+                # Low battery - let's check half as often
+                interval = interval * 2
+
+            intervals[device.name] = interval
+
+        return max(
+            int(min(intervals.items(), key=operator.itemgetter(1))[1]),
+            self._max_interval,
+        )
+
+    def keep_alive(self, now):
+        """Keep the API alive."""
+        if self.api is None:
+            self.reset_account()
+
+        if self.api is None:
+            return
+
+        self.api.authenticate()
+        self.update_devices()
+
+    @property
+    def name(self):
+        """Return the account name."""
+        return self._accountname
+
+
+class IcloudDevice(Entity):
+    """Representation of a iCloud device."""
+
+    def __init__(self, account: IcloudAccount, device: AppleDevice):
+        """Initialize the iCloud device."""
+        self.hass = account.hass
+        self._account = account
+        self._accountname = account.name
+        self._accountname_slug = slugify(self._accountname.partition("@")[0])
+
+        self._device = device
+        self._status = device.status(DEVICE_STATUS_SET)
```

Comment:
What happens in this call? Any I/O?

Reply:
Retrieved the dict device status https://github.com/picklepete/pyicloud/blob/9bb6d750662ce24c8febc94807ddbdcdf3cadaa2/pyicloud/services/findmyiphone.py#L115, but oooops yes, also calls `refresh_client` wich calls the API to `ensures that the location data is up-to-date`.

What shoul I do to eliminate this call in the `IcloudDevice.__init__` ?
I think I'm just gonna pass it to the params

Reply:
Yeah, we can pass the status to the instance after it has been created. Or have a dedicated setup method on the instance.